# Description

To ensure the predictions we decided to  supplement our EEG Machine Learning models with a CNN model based on the two different classes given by the writing samples.

# Importing the Libraries

In [1]:
import tensorflow as tf
from keras.preprocessing.image import ImageDataGenerator

# Data Preprocessing

## Preprocessing the Training Set

The input image is on RGB. Every image is made up of pixels that range from 0 to 255. We need to normalize them i.e convert the range between 0 to 1 before passing it to the model.

In [2]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)
training_set = train_datagen.flow_from_directory('Dataset/Training_Set',
                                                 target_size = (64, 64),
                                                 batch_size = 32,
                                                 class_mode = 'binary')

Found 146 images belonging to 2 classes.


The total of images in the training set is given by 73 * 2 = 146. 

## Preprocessing the Test Set

In [3]:
test_datagen = ImageDataGenerator(rescale = 1./255)
test_set = test_datagen.flow_from_directory('Dataset/Test_Set',
                                            target_size = (64, 64),
                                            batch_size = 32,
                                            class_mode = 'binary')

Found 34 images belonging to 2 classes.


The total of images in the test set is given by 16 * 2 = 34. 

# Building the CNN Model

## Initializing the Model

In [4]:
Model = tf.keras.models.Sequential()

## Adding First Convolution Layer 

The convolution layer is the layer where the filter is applied to our input image to extract or detect its features. A filter is applied to the image multiple times and creates a feature map which helps in classifying the input image.

In [5]:
Model.add(tf.keras.layers.Conv2D(filters=32, kernel_size=3, activation='relu', input_shape=[64, 64, 3]))

## Pooling the First Layer

The pooling layer is applied after the Convolutional layer and is used to reduce the dimensions of the feature map which helps in preserving the important information or features of the input image and reduces the computation time.

The pooling operation involves sliding a two-dimensional filter over each channel of feature map and summarising the features lying within the region covered by the filter. 

In [6]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))

## Adding a Second Convolutional Layer

In [7]:
Model.add(tf.keras.layers.Conv2D(filters=32, kernel_size=3, activation='relu'))

## Pooling the Second Layer

In [8]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))

## Flattening

The flattening step involves taking the pooled feature map that is generated in the pooling step and transforming it into a one-dimensional vector. This vector will now become the input layer of an artificial neural network.

In [9]:
Model.add(tf.keras.layers.Flatten())

## Full Connection

The full connection step involves chaining an artificial neural network onto our existing convolutional neural network.

In [10]:
Model.add(tf.keras.layers.Dense(units=128, activation='relu'))

## Output Layer

We add the output layer. In this project we must have units=1 because we need to classify 2 different classes.

In [11]:
Model.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Compiling the CNN

In [12]:
Model.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

## Training the CNN

In [13]:
Model.fit(x = training_set, validation_data = test_set, epochs = 50)

Epoch 1/50
5/5 [==============================] - 7s 1s/step - loss: 0.8415 - accuracy: 0.5137 - val_loss: 0.6979 - val_accuracy: 0.5000
Epoch 2/50
5/5 [==============================] - 4s 885ms/step - loss: 0.6966 - accuracy: 0.5000 - val_loss: 0.6958 - val_accuracy: 0.5000
Epoch 3/50
5/5 [==============================] - 5s 1s/step - loss: 0.6956 - accuracy: 0.5000 - val_loss: 0.6930 - val_accuracy: 0.5000
Epoch 4/50
5/5 [==============================] - 4s 987ms/step - loss: 0.6942 - accuracy: 0.4863 - val_loss: 0.6927 - val_accuracy: 0.5000
Epoch 5/50
5/5 [==============================] - 4s 870ms/step - loss: 0.6938 - accuracy: 0.5000 - val_loss: 0.6929 - val_accuracy: 0.5000
Epoch 6/50
5/5 [==============================] - 4s 884ms/step - loss: 0.6930 - accuracy: 0.5000 - val_loss: 0.6927 - val_accuracy: 0.5000
Epoch 7/50
5/5 [==============================] - 5s 914ms/step - loss: 0.6936 - accuracy: 0.4589 - val_loss: 0.6929 - val_accuracy: 0.5000
Epoch 8/50
5/5 [==========

## Evaluating the Model

Firstly we evaluate the ability of the model in predicting our training set. 

In [14]:
acc_training = Model.evaluate(training_set)
print ("The accurancy of the model on the training set is", round(acc_training[1]*100, 2), "%")

5/5 [==============================] - 3s 670ms/step - loss: 0.5492 - accuracy: 0.7397
The accurancy of the model on the training set is 73.97 %


In [15]:
acc_test = Model.evaluate(test_set)
print ("The accurancy of the model on the test set is", round(acc_test[1]*100, 2), "%")

2/2 [==============================] - 1s 45ms/step - loss: 0.5869 - accuracy: 0.6765
The accurancy of the model on the test set is 67.65 %
